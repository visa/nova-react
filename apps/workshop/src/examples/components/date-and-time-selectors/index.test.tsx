/**
 *              Â© 2025 Visa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
 
import { fireEvent, render, screen } from '@testing-library/react';
import { axe } from 'jest-axe';

import metaData from './meta.json';

import { DefaultDateSelector } from './default-date-selector';
import { ReadOnlyDateSelector } from './read-only-date-selector';
import { DisabledDateSelector } from './disabled-date-selector';
import { DateSelectorWithError } from './date-selector-with-error';
import { MinMaxDateInput } from './min-max-date-input';
import { DefaultDateRangeSelector } from './default-date-range-selector';
import { StackedDateRangeSelector } from './stacked-date-range-selector';
import { DefaultTimeSelector } from './default-time-selector';
import { ReadOnlyTimeSelector } from './read-only-time-selector';
import { DisabledTimeSelector } from './disabled-time-selector';
import { TimeSelectorWithError } from './time-selector-with-error';

const examples = [
  {
    Component: DefaultDateSelector,
    title: metaData['default-date-selector'].title,
  },
  {
    Component: ReadOnlyDateSelector,
    title: metaData['read-only-date-selector'].title,
  },
  {
    Component: DisabledDateSelector,
    title: metaData['disabled-date-selector'].title,
  },
  {
    Component: DateSelectorWithError,
    title: metaData['date-selector-with-error'].title,
  },
  {
    Component: MinMaxDateInput,
    title: metaData['min-max-date-input'].title,
  },
  {
    Component: DefaultDateRangeSelector,
    title: metaData['default-date-range-selector'].title,
  },
  {
    Component: StackedDateRangeSelector,
    title: metaData['stacked-date-range-selector'].title,
  },
  {
    Component: DefaultTimeSelector,
    title: metaData['default-time-selector'].title,
  },
  {
    Component: ReadOnlyTimeSelector,
    title: metaData['read-only-time-selector'].title,
  },
  {
    Component: DisabledTimeSelector,
    title: metaData['disabled-time-selector'].title,
  },
  {
    Component: TimeSelectorWithError,
    title: metaData['time-selector-with-error'].title,
  },
];

describe('Input examples', () => {
  examples.forEach(({ Component, title }) => {
    it(`${title} should render correctly`, async () => {
      const { container } = render(<Component />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
      expect(container).toMatchSnapshot();
    });
  });

  describe('date range', () => {
    it('should adjust ranges after changing end and start dates', () => {
      render(<DefaultDateRangeSelector />);
      const startDateInput = screen.getByLabelText<HTMLInputElement>('Start Date Label (required)');
      const endDateInput = screen.getByLabelText<HTMLInputElement>('End Date Label (required)')!;
      fireEvent.change(startDateInput, { target: { value: '2022-01-01' } });
      fireEvent.change(endDateInput, { target: { value: '2022-01-31' } });
      expect(startDateInput.max).toBe('2022-01-31');
      expect(endDateInput.min).toBe('2022-01-01');
    });
  });

  describe('stacked date range', () => {
    it('should adjust ranges after changing end and start dates', () => {
      render(<StackedDateRangeSelector />);
      const startDateInput = screen.getByLabelText<HTMLInputElement>('Start Date Label (required)');
      const endDateInput = screen.getByLabelText<HTMLInputElement>('End Date Label (required)')!;
      fireEvent.change(startDateInput, { target: { value: '2022-01-01' } });
      fireEvent.change(endDateInput, { target: { value: '2022-01-31' } });
      expect(startDateInput.max).toBe('2022-01-31');
      expect(endDateInput.min).toBe('2022-01-01');
    });
  });

  describe('date selector with error', () => {
    it('should show error', () => {
      const { container } = render(<DateSelectorWithError />);
      const errorButton = container.querySelector<HTMLButtonElement>('#date-selector-with-error-submit-button')!;
      fireEvent.click(errorButton);
      expect(screen.getByText('This is required text that describes the error in more detail.')).toBeInTheDocument();
      const resetButton = container.querySelector<HTMLButtonElement>('#date-selector-with-error-reset-button')!;
      fireEvent.click(resetButton);
      expect(screen.queryByText('This is required text that describes the error in more detail.')).toBeNull();
    });
  });
});

describe('time selector with error', () => {
  it('should show error', () => {
    const { container } = render(<TimeSelectorWithError />);
    const errorButton = container.querySelector<HTMLButtonElement>('#time-selector-with-error-submit-button')!;
    fireEvent.click(errorButton);
    expect(screen.getByText('This is required text that describes the error in more detail.')).toBeInTheDocument();
    const resetButton = container.querySelector<HTMLButtonElement>('#time-selector-with-error-reset-button')!;
    fireEvent.click(resetButton);
    expect(screen.queryByText('This is required text that describes the error in more detail.')).toBeNull();
  });
});