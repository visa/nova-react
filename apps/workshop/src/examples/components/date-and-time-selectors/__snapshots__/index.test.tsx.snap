// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input examples Date selector with disabled dates should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-flex-wrap v-gap-4"
  >
    <label
      class="v-label"
      for="min-max-date"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        id="min-max-date"
        max="1999-06-29"
        min="1999-01-27"
        required=""
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Date selector with error should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-gap-4"
  >
    <label
      class="v-label"
      for="date-selector-with-error"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        aria-invalid="false"
        class="v-input"
        id="date-selector-with-error"
        required=""
        type="date"
      />
    </div>
  </div>
  <div
    class="v-flex v-gap-12 v-mt-16"
  >
    <button
      class="v-button"
      id="date-selector-with-error-submit-button"
    >
      Submit
    </button>
    <button
      class="v-button v-button-secondary"
      id="date-selector-with-error-reset-button"
    >
      Reset
    </button>
  </div>
</div>
`;

exports[`Input examples Default date range selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-wrap v-gap-12"
  >
    <div
      class="v-flex v-flex-col v-flex-grow v-gap-4"
    >
      <label
        class="v-label"
        for="default-date-range-selector-start-date"
      >
        Start Date Label (required)
      </label>
      <div
        class="v-surface v-input-container"
      >
        <input
          class="v-input"
          id="default-date-range-selector-start-date"
          required=""
          type="date"
        />
      </div>
    </div>
    <div
      class="v-flex v-flex-col v-flex-grow v-gap-4"
    >
      <label
        class="v-label"
        for="default-date-range-selector-end-date"
      >
        End Date Label (required)
      </label>
      <div
        class="v-surface v-input-container"
      >
        <input
          class="v-input"
          id="default-date-range-selector-end-date"
          required=""
          type="date"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input examples Default date selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-gap-4"
  >
    <label
      class="v-label"
      for="default-date-selector"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        id="default-date-selector"
        required=""
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Default time selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-flex-grow v-gap-4"
  >
    <label
      class="v-label"
      for="default-time-selector-time"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        id="default-time-selector-time"
        required=""
        type="time"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Disabled date selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-gap-4"
  >
    <label
      class="v-label"
      for="disabled-date-selector"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        disabled=""
        id="disabled-date-selector"
        required=""
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Disabled time selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-flex-grow v-gap-4"
  >
    <label
      class="v-label"
      for="disabled-time-selector-time"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        disabled=""
        id="disabled-time-selector-time"
        required=""
        type="time"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Read-only date selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-gap-4"
  >
    <label
      class="v-label"
      for="read-only-date-selector"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        id="read-only-date-selector"
        readonly=""
        required=""
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Read-only time selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-flex-grow v-gap-4"
  >
    <label
      class="v-label"
      for="read-only-time-selector-time"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        class="v-input"
        id="read-only-time-selector-time"
        readonly=""
        required=""
        type="time"
      />
    </div>
  </div>
</div>
`;

exports[`Input examples Stacked date range selector should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-gap-12"
  >
    <div
      class="v-flex v-flex-col v-flex-grow v-gap-4"
    >
      <label
        class="v-label"
        for="stacked-date-range-selector-start-date"
      >
        Start Date Label (required)
      </label>
      <div
        class="v-surface v-input-container"
      >
        <input
          class="v-input"
          id="stacked-date-range-selector-start-date"
          required=""
          type="date"
        />
      </div>
    </div>
    <div
      class="v-flex v-flex-col v-flex-grow v-gap-4"
    >
      <label
        class="v-label"
        for="stacked-date-range-selector-end-date"
      >
        End Date Label (required)
      </label>
      <div
        class="v-surface v-input-container"
      >
        <input
          class="v-input"
          id="stacked-date-range-selector-end-date"
          required=""
          type="date"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Input examples Time selector with error should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-flex-grow v-gap-4"
  >
    <label
      class="v-label"
      for="time-selector-with-error-time"
    >
      Label (required)
    </label>
    <div
      class="v-surface v-input-container"
    >
      <input
        aria-invalid="false"
        class="v-input"
        id="time-selector-with-error-time"
        required=""
        type="time"
      />
    </div>
  </div>
  <div
    class="v-flex v-gap-12 v-mt-16"
  >
    <button
      class="v-button"
      id="time-selector-with-error-submit-button"
    >
      Submit
    </button>
    <button
      class="v-button v-button-secondary"
      id="time-selector-with-error-reset-button"
    >
      Reset
    </button>
  </div>
</div>
`;
